### PR TITLE
`ias` replacement

### PR DESCRIPTION
**Pull request by raoulstrackx**

### CODE REVIEW

File: intel-sgx/ias/src/client.rs [hunk 1]
```diff
@@ -355,30 +331,31 @@ impl Client {
...
+        let report: VerifyAttestationEvidenceResponse<Unverified> = VerifyAttestationEvidenceResponse::deserialize(deser)
+            .map_err(|e| format!("Error deserializing JSON response: {}", e))?;
+
+        debug!("IAS verification report: {:?}", report);
 
-        // Prior to API documentation v6.0, advisory_url and advisory_ids
-        // were returned in headers instead of the report itself
-        match (&advisory_url, &report.advisory_url) {
+        // since API documentation v6.0 advisory_url and advisory_ids
+        // are returned in the body of the response
```
Comment (raoulstrackx) [inline intel-sgx/ias/src/client.rs:340]: Is there a good reason these were still in `IasResult`? If not, we can remove them and there is also no longer a need to parse the `VerifyAttestationEvidenceResponse` in this function

File: intel-sgx/ias/src/client.rs [hunk 2]
```diff
@@ -64,36 +60,20 @@ pub struct IasVerificationResult {
...
-        } = res;
-        AttestationEmbeddedIasReport::from(
-            sgx_pkix::attestation::AttestationEmbeddedIasReport {
-                http_body: raw_report.into(),
-                report_sig: signature.into(),
-                certificates: cert_chain.into_iter().map(Into::into).collect()
-            }
-        )
+impl IasVerificationResult {
+    pub fn report<C: Crypto>(&self, ca_certificates: &[&[u8]]) -> Result<VerifyAttestationEvidenceResponse> {
```
Comment (jethrogb) [inline intel-sgx/ias/src/client.rs:64]: `fn verify` instead? Also add a doc comment what you're actually verifying

File: intel-sgx/ias/src/client.rs (hunk 1 above)
Comment (jethrogb) [reply intel-sgx/ias/src/client.rs:340]: These are here to be able to offer a consistent Rust API regardless of REST API version?

Comment (raoulstrackx) [reply intel-sgx/ias/src/client.rs:340]: If this crate was already publicly available before v6.0, I could understand but this is the initial public version. Do we expect IAS to go back to a previous version? Or are we using a proxy that may not use v6.0?

File: intel-sgx/ias/src/api.rs [hunk 3]
```diff
@@ -148,78 +167,44 @@ fn less_than_three(&v: &u64) -> bool {
     v < 3
 }
 
-// Intel security advisory ids are strings of the form "INTEL-SA-ddddd".
-// https://www.intel.com/content/www/us/en/security-center/default.html has more details.
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
-pub struct IasAdvisoryId(String);
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct IasAdvisoryId(pub String);
```
[the remark this answers is not in the document]
Comment (raoulstrackx) [reply intel-sgx/ias/src/api.rs:171]: Yes that's better. Fixed.

File: intel-sgx/ias/src/client.rs (hunk 2 above)
Comment (raoulstrackx) [reply intel-sgx/ias/src/client.rs:64]: Fixed

File: intel-sgx/ias/examples/attestation-tool.rs [hunk 4]
```diff
@@ -152,44 +136,33 @@ async fn main() {
...
-                Ok(report) => report,
-                Err(err) => {
-                    println!("IAS verification error:");
-                    println!("  {}", err);
-                    return;
-                }
-            };
+            let report = response
+                .verify::<Mbedtls>(&[TEST_REPORT_SIGNING_CERT.as_slice()])
+                .expect("Corrupt report");
```
Comment (jethrogb) [inline intel-sgx/ias/examples/attestation-tool.rs:141]: Don't expect. Just print an error message and quit. Also, the error message should be "Unable to verify signature on IAS report".

File: intel-sgx/ias/src/api.rs [hunk 5]
```diff
@@ -230,46 +204,46 @@ impl SafeToDeserializeInto for Unverified {}
...
     #[serde(skip_serializing_if = "Option::is_none", rename = "advisoryIDs")]
-    pub(crate) advisory_ids: Option<Vec<IasAdvisoryId>>,
+    advisory_ids: Option<Vec<IasAdvisoryId>>,
 
     #[serde(skip)]
-    pub(crate) type_: PhantomData<V>
+    type_: PhantomData<V>
 }
 
 impl<V: VerificationType> VerifyAttestationEvidenceResponse<V> {
```
Comment (jethrogb) [inline intel-sgx/ias/src/api.rs:249]: Why are these methods available unverified? Is this for the IasVerificationResult? Then the unverified version should be `pub(crate)`.

Comment (raoulstrackx) [reply intel-sgx/ias/src/api.rs]: Yes that's solely for the `IasVerificationResult`. Fixed

File: intel-sgx/ias/examples/attestation-tool.rs (hunk 4 above)
Comment (raoulstrackx) [reply intel-sgx/ias/examples/attestation-tool.rs]: Fixed

File: intel-sgx/ias/examples/attestation-tool.rs [hunk 6]
```diff
@@ -27,26 +24,13 @@ const IAS_DEV_OLD_URL: &'static str = "https://test-as.sgx.trustedservices.intel
...
-    }
-}
-
-impl PlatformVerifier for IgnorePlatformState {
-    fn verify(&self, _for_self: bool, _nonce: &Option<String>, _isv_enclave_quote_status: QuoteStatus, _advisories: &Vec<IasAdvisoryId>) -> Result<(), Error> {
-        Ok(())
-    }
+lazy_static::lazy_static!{
+    // This is the IAS report signing certificate.
+    static ref TEST_REPORT_SIGNING_CERT: Vec<u8> =
```
[the remark this answers is not in the document]
Comment (raoulstrackx) [reply intel-sgx/ias/examples/attestation-tool.rs]: Fixed

File: .travis.yml [hunk 7]
```diff
@@ -38,10 +38,10 @@ matrix:
       script:
         - cargo test --verbose --locked --all --exclude sgxs-loaders && [ "$(echo $(nm -D target/debug/sgx-detect|grep __vdso_sgx_enter_enclave))" = "w __vdso_sgx_enter_enclave" ]
         - cargo test --verbose --locked -p sgxs-tools --features pe2sgxs --bin isgx-pe2sgx
-        - cargo test --verbose --locked -p ias --features mbedtls
-        - cargo test --verbose --locked -p ias --features "client, mbedtls"
         - cargo test --verbose --locked -p dcap-ql --features link
         - cargo test --verbose --locked -p dcap-ql --features verify
+        - cargo test --verbose --locked -p ias --features mbedtls
+        - cargo test --verbose --locked -p ias --features "client, mbedtls"
```
[the remark this answers is not in the document]
Comment (raoulstrackx) [reply .travis.yml]: Fixed

File: intel-sgx/ias/src/verifier.rs [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+                f.write_fmt(format_args!("enclave peer miscselect mismatch, masked local = {:?}, masked remote = {:?}",
+                                         local, remote)),
+        }
+    }
+}
+
+/// Verify that `report` is correctly signed by a key that chains to one of the
+/// trusted certificates in `ca_certificates`.
+///
+/// Does NOT verify the report contents.
```
Comment (jethrogb) [inline intel-sgx/ias/src/verifier.rs:123]: Doc comment should be on the public function

Comment (raoulstrackx) [reply intel-sgx/ias/src/verifier.rs]: Fixed

**Pull request merged**